src/main/java/org/truffleruby/core/array/ArrayStrategy.java

```diff
@@ -97,6 +97,10 @@ public ArrayStrategy generalize(ArrayStrategy other) {
             return other.generalizeForMutation();
         }
 
+        if (other instanceof IntToObjectGeneralizationArrayStrategy) {
+            return other.generalize(this);
+        }
```

Comment:
IntToObjectGeneralizationArrayStrategy defines `generalize` and doesn't call this, so I don't think this will work.
I think IntToObjectGeneralizationArrayStrategy.generalize() should just return an Object[] strategy when generalizing against a strategy different than itself.